src/test/java/com/arpnetworking/metrics/common/sources/KafkaSourceTest.java

```diff
@@ -0,0 +1,173 @@
+import com.arpnetworking.steno.LogBuilder;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.test.StringParser;
+import com.google.common.base.Charsets;
+import com.google.common.collect.Maps;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.MockConsumer;
+import org.apache.kafka.clients.consumer.OffsetResetStrategy;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Unit tests for the <code>KafkaSource</code> class.
+ *
+ * @author Joey Jackson (jjackson at dropbox dot com)
+ */
+public class KafkaSourceTest {
+    private KafkaSource<String, String> _source;
+    private static final List<String> EXPECTED = Arrays.asList("value0", "value1", "value2");
+    private static final String TOPIC = "test_topic";
+    private static final int PARTITION = 0;
+    private static final int POLL_TIME_MILLIS = 1;
+    private static final int TIMEOUT = 1000;
+    private Logger _logger;
+    private LogBuilder _logBuilder;
+
+    @Before
+    public void setUp() {
+        _logger = Mockito.mock(Logger.class);
+        _logBuilder = Mockito.mock(LogBuilder.class);
+        Mockito.when(_logger.trace()).thenReturn(_logBuilder);
+        Mockito.when(_logger.debug()).thenReturn(_logBuilder);
+        Mockito.when(_logger.info()).thenReturn(_logBuilder);
+        Mockito.when(_logger.warn()).thenReturn(_logBuilder);
+        Mockito.when(_logger.error()).thenReturn(_logBuilder);
+        Mockito.when(_logBuilder.setMessage(Mockito.anyString())).thenReturn(_logBuilder);
+        Mockito.when(_logBuilder.addData(Mockito.anyString(), Mockito.any())).thenReturn(_logBuilder);
+        Mockito.when(_logBuilder.addContext(Mockito.anyString(), Mockito.any())).thenReturn(_logBuilder);
+        Mockito.when(_logBuilder.setEvent(Mockito.anyString())).thenReturn(_logBuilder);
+        Mockito.when(_logBuilder.setThrowable(Mockito.any(Throwable.class))).thenReturn(_logBuilder);
+    }
+
+    @Test
+    public void testSourceSuccess() {
+        createHealthySource();
+        final Observer observer = Mockito.mock(Observer.class);
+        _source.attach(observer);
+        _source.start();
+        for (String expected : EXPECTED) {
+            Mockito.verify(observer, Mockito.timeout(TIMEOUT)).notify(_source, expected);
+        }
+        _source.stop();
+    }
+
+    @Test
+    public void testSourceKafkaException() {
+        createExceptionSource(KafkaException.class);
+        _source.start();
+        Mockito.verify(_logBuilder, Mockito.timeout(TIMEOUT)).setMessage("Consumer received Kafka Exception");
```

Comment:
Testing via log message is generally a really bad idea. Not only are the tests brittle, they are dependent on the log message actually indicating the behavior/data you want to assert occurred (e.g. a byproduct instead of a direct observation).

Reply:
I agree in general, but in this case isnt logging the actual direct desired behavior in the case of an error (besides the thread sleeping)?